src/org/sosy_lab/java_smt/solvers/z3/Z3FormulaCreator.java

```diff
@@ -386,16 +387,53 @@ public Long getFloatingPointType(FormulaType.FloatingPointType type) {
     return fpSort;
   }
 
+  private static class Z3AstReference extends PhantomReference<Z3Formula> {
+    private long z3Ast;
+    private @Nullable Z3AstReference prev;
+    private @Nullable Z3AstReference next;
+
+    // To generate dummy head and tail nodes
+    public Z3AstReference() {
+      super(null, null);
+    }
+
+    public Z3AstReference(Z3Formula referent, ReferenceQueue<? super Z3Formula> q, long z3Ast) {
+      super(referent, q);
+      this.z3Ast = z3Ast;
+    }
+
+    public void insert(Z3AstReference ref) {
+      assert next != null;
+      ref.prev = this;
+      ref.next = this.next;
+      ref.next.prev = ref;
+      this.next = ref;
+    }
+
+    public void cleanup(Long environment) {
```

Comment:
Using `Long` here and in `storePhantomReference` doesn't really provide a benefit. The current callers have `Long`s anyway so at least it doesn't cause additional boxing, but both methods only unbox the `Long`, so accepting `long`  works the same and helps to avoid confusion and accidental boxing the future.

Reply:
I agree.